pkg/controller/api.go

```diff
@@ -67,39 +74,109 @@ func NewAPICallbacks(m ctrl.Manager, of xpresource.ManagedKind) *APICallbacks {
 		tr.SetConditions(resource.AsyncOperationFinishedCondition())
-		return errors.Wrap(ac.kube.Status().Update(ctx, tr), errStatusUpdate)
+		return errors.Wrapf(ac.kube.Status().Update(ctx, tr), errUpdateStatusFmt, tr.GetObjectKind().GroupVersionKind().String(), name, op)
 	}
 }
 
-// Destroy makes sure the error is saved in async operation condition.
-func (ac *APICallbacks) Destroy(name string) terraform.CallbackFn {
+// Create makes sure the error is saved in async operation condition.
+func (ac *APICallbacks) Create(name string) terraform.CallbackFn {
 	return func(err error, ctx context.Context) error {
-		nn := types.NamespacedName{Name: name}
-		tr := ac.newTerraformed()
-		if kErr := ac.kube.Get(ctx, nn, tr); kErr != nil {
-			return errors.Wrap(kErr, errGet)
+		cErr := ac.callbackFn(name, "create")(err, ctx)
+		switch {
+		case err != nil:
+			ac.EventHandler.requestReconcile(name)
+		default:
+			ac.EventHandler.rateLimiter.Forget(reconcile.Request{
+				NamespacedName: types.NamespacedName{
+					Name: name,
+				},
+			})
 		}
-		tr.SetConditions(resource.LastAsyncOperationCondition(err))
-		tr.SetConditions(resource.AsyncOperationFinishedCondition())
-		return errors.Wrap(ac.kube.Status().Update(ctx, tr), errStatusUpdate)
+		return cErr
 	}
 }
+
+// Update makes sure the error is saved in async operation condition.
+func (ac *APICallbacks) Update(name string) terraform.CallbackFn {
+	return func(err error, ctx context.Context) error {
+		cErr := ac.callbackFn(name, "update")(err, ctx)
+		switch {
+		case err != nil:
+			ac.EventHandler.requestReconcile(name)
+		default:
+			ac.EventHandler.rateLimiter.Forget(reconcile.Request{
+				NamespacedName: types.NamespacedName{
+					Name: name,
+				},
+			})
+		}
+		return cErr
+	}
+}
+
+// Destroy makes sure the error is saved in async operation condition.
+func (ac *APICallbacks) Destroy(name string) terraform.CallbackFn {
+	return ac.callbackFn(name, "destroy")
+}
+
+type eventHandler struct {
+	queue       workqueue.RateLimitingInterface
+	rateLimiter workqueue.RateLimiter
+	mu          *sync.Mutex
+}
+
+func (e *eventHandler) requestReconcile(name string) bool {
+	e.mu.Lock()
+	defer e.mu.Unlock()
+	if e.queue == nil {
+		return false
+	}
+	item := reconcile.Request{
+		NamespacedName: types.NamespacedName{
+			Name: name,
+		},
+	}
+	e.queue.AddAfter(item, e.rateLimiter.When(item))
+	return true
+}
+
+func (e *eventHandler) Create(_ context.Context, _ event.CreateEvent, limitingInterface workqueue.RateLimitingInterface) {
+	e.mu.Lock()
+	defer e.mu.Unlock()
+	if e.queue == nil {
+		e.queue = limitingInterface
+	}
+}
+
+func (e *eventHandler) Update(_ context.Context, _ event.UpdateEvent, _ workqueue.RateLimitingInterface) {
+}
+
+func (e *eventHandler) Delete(_ context.Context, _ event.DeleteEvent, _ workqueue.RateLimitingInterface) {
+}
+
+func (e *eventHandler) Generic(_ context.Context, _ event.GenericEvent, _ workqueue.RateLimitingInterface) {
+}
```

Comment:
As we discussed offline, we may consider copying the same logic in the `(e *eventHandler) Create` to other interface implementations.

Reply:
Agreed. The previous assumption is that event handler's `Create` will always be called for the controller. But we don't know if this is true. Let's assume less and do some more synchronization.